### DISCUSSION TITLE
Lifecycle API: rebuilding the View/Widget tree

### DISCUSSION THREAD

**Discussion by ignatz** [lapce/floem#426] [Q&A] [answered]
Hi floemers,

I was looking for the means to rebuild the contents contents of an entire window or the entire application.  Looking through the APIs I noticed that:

* most of the "inject event" apis are crate private but even than I couldn't find an appropriate lifecycle event
* that the `view_fn` view builder is of type `FnOnce`, so rebuilding the view is probably out of the question.
* and the only widget building I could find is the call to `view.build()` in `WindowHandle::new`.

So, I was wondering if it would be possible to get or add a hook for either:

* rebinding an existing window to a new view (I can already close a window and open a new one)
* or an accessible lifecycle event for rebuilding the widget tree

?

For context, I'm trying to build hot reload/restart capabilities.

Thanks!

**Comment by dzhou121** [comment, marked as answer]
The easiest way is probably to use `dyn_container` at the root. Then you can swap what's inside the `dyn_container` as you like.

**Reply by ignatz** [reply]
@dzhou121 et al, I did play a little more and noticed that a `dyn_container::update` will dispose the child scope and thus all associated state. 

This works great for something like a hot-restart, i.e build everything fresh, however I'd like to rebuild the view/widgets while preserving the state. Is that something that would be possible?

**Reply by dzhou121** [reply]
You can define your state in the parent, and pass the signals to the child. And because those signals are defined in the parent, they are under the scope of parent as well,

**Reply by ignatz** [reply]
My setup unfortunately doesn't allow for that. I'm wrapping an application given to me, i.e. I'm not owning the state. I'd like to rebuild the window contents, assuming the build function may have changed while preserving the state. Sorry, this must feel very snowflake. Happy to explain more

**Reply by dzhou121** [reply]
I don't quite get this bit "I'm not owning the state"

**Reply by ignatz** [reply]
Imagine you wrapped floem for some scripting language. You're being given a build method very similar `floem::Application::window()`'s app_view argument `Fn(...) -> View`. But because you're going in and out between languages, there isn't really a good way to capture or pass much state besides some pointers.

So far so good, I've wrapped a couple of widgets and can nicely build views and widget trees. Now, the user changes view/widget building code, the runtime re-interprets/compiles/jits/whatever the code and I can rebuild the view using the dyn_container. All of this works like a charm. In an ideal world however, I'd like to keep the state in tact.
For example, in many hot-reloadable frameworks when tweaking the pixel position of a widget you'd expect the widget to move on-file-save but not the state to reset (that would be hot restart).

Now this is where it gets tricky: since you're rerunning the builder, you're also reconstructing child views, widgets, signals, ... , which means you'd also have to cleanup previous subscriptions, effects, ... and re-associate the state. After thinking about it, I'm not even sure that it makes sense w/o stable ids. I had a quick look at SolidJS and at least out-of-the box in dev mode it does simply reset state :/.

FWIW, frameworks w/o signal and virtual dom diffing do this all the time (React, Flutter, ...). Since they rebuild the component tree on every frame, they have to inject state on every build.rebuild.
In practice, these frameworks also rely on you to sometimes assign unique keys/ids for disambiguation. I guess signals would have similar needs, if I was going to re-associate state.

At this point, I've basically convinced myself (hopefully wrongly so) that, this isn't currently feasible with floem but I'd love to hear your input. I can also understand that this is probably not super interesting to you working entirely in a language that isn't very hot reloadable... Idk. Maybe you've considered scripting or an external styling language akin to CSS :woman_shrugging:.

**Reply by dzhou121** [reply]
This is a very interesting subject. I've been thinking about how or is it even possible to do hot reloading in Floem. 

It's kind of no difference in hot reload from a different language or hot reload from your IDE, because it's going to be cross process communications anyways. 

I don't know how you instruct Floem to build the view tree from the scripting language (it's super cool by the way), but do you think it's possible to have some sort of diffing of the view tree locally from the scripting language, and then instead of swapping everything out from the root view, but actually swapping only the changing ones?

**Reply by ignatz** [reply]
Glad to hear you're interested. Much appreciated.

> It's kind of no difference in hot reload from a different language or hot reload from your IDE, because it's going to be cross process communications anyways.

You're probably thinking about how to trigger the hot reload. Agreed, there's some external signal whether it's a file watcher or a button in an IDE, all the same. FWIW, I was more talking about passing state across language boundaries and across lexicographical scopes when the UI is described in a different language. For this part at least, there are no process boundaries. 

> I don't know how you instruct Floem to build the view tree from the scripting language (it's super cool by the way)

I kept it straight forward by simply wrapping construction/destruction of widgets and views. Imagine a C-API like:

```C
Widget* new_floem_text(const char* text);
void destroy_widget(Widget*);
```

>  but do you think it's possible to have some sort of diffing of the view tree locally from the scripting language, and then instead of swapping everything out from the root view, but actually swapping only the changing ones?

That's an interesting idea. I guess that would be quite hard to do, since I don't really know how a change to file x line y maps to the widget tree. It could affect one, all, or none.

I ended up trying something utterly simple, which sees to work alright for now... I added an optional argument for providing an explicit "key". For signals with keys,  I'm keeping a copy of the state around and when reconstructing the view, including the signals, I inject the persisted state rather than the initial state. It's easy because it didn't require any changes to floem but one could imagine lowering this approach into floem if you folks were thinking about experimenting with e.g. hot reloadable shared objects.

I'm happy to share the work once things get into a better state. In any case, I can't thank you enough @dzhou121 , interacting has been a real treat :pray:

**Reply by dzhou121** [reply]
I think your approach to persist the state is actually a very clean way, since for a declarative UI, you would expect the same state would result in the same view tree.

It's definitely something we want to bring into Floem, which will be useful for your use case (it's going to be powerful if you can bind Floem to Python or any other scripting languages). Hot reload can probably be implemented in this way. And now thinking about it, the idea "same state result in the same view tree" can shape how you can do UI testing in Floem. 

Thank you @ignatz for your work and I would look forward to seeing it.

**Reply by ignatz** [reply]
It's very basic but I hope it gets the idea across:

[Screencast from 2024-05-04 19-41-22.webm](https://github.com/lapce/floem/assets/111986/f205fc5a-a3ee-4431-aa74-824833724e7b)

EDIT: 

>  And now thinking about it, the idea "same state result in the same view tree" can shape how you can do UI testing in Floem.

Hadn't thought that far but that's a great point.

**Reply by dzhou121** [reply]
That is utterly cool. Thanks for exploring it and sharing it! 

I'll see if I can bring it into Floem after my ECS PR.

**Comment by ignatz** [comment]
@dzhou121 that's awesome, this gets me 90% there. I was able to add a button to the UI to rebuild the entire widget tree.

Do you have a suggestion for me on how to drive it externally? Signals under the hood deep on the thread-local `RUNTIME`, so I cannot simply pass it to  a different thread and Application::run takes over the main thread for running the render loop. Is there either:
 * a way for me to periodically poll a buffered channel on the UI thread, e.g. register a callback that gets called for every frame or periodic timer intervals?
 * or is there a signal implementation where at least the writer-end can be sent to a different thread?

Thanks!

**Reply by dzhou121** [reply]
Yes, that would be `create_ext_action`

https://docs.rs/floem/latest/floem/ext_event/fn.create_ext_action.html

**Reply by dzhou121** [reply]
Or `create_signal_from_channel`

https://docs.rs/floem/latest/floem/ext_event/fn.create_signal_from_channel.html

**Reply by ignatz** [reply]
Awesome. That did the trick :bowing_woman:
